polkadot/runtime/parachains/src/paras/mod.rs

```diff
@@ -1160,59 +1188,149 @@ pub mod pallet {
+					let current_session = shared::CurrentSessionIndex::<T>::get();
+					if stmt.session_index < current_session {
+						return InvalidTransaction::Stale.into()
+					} else if stmt.session_index > current_session {
+						return InvalidTransaction::Future.into()
+					}
 
-			let current_session = shared::CurrentSessionIndex::<T>::get();
-			if stmt.session_index < current_session {
-				return InvalidTransaction::Stale.into()
-			} else if stmt.session_index > current_session {
-				return InvalidTransaction::Future.into()
-			}
+					let validator_index = stmt.validator_index.0 as usize;
+					let validators = shared::ActiveValidatorKeys::<T>::get();
+					let validator_public = match validators.get(validator_index) {
+						Some(pk) => pk,
+						None =>
+							return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
+					};
 
-			let validator_index = stmt.validator_index.0 as usize;
-			let validators = shared::ActiveValidatorKeys::<T>::get();
-			let validator_public = match validators.get(validator_index) {
-				Some(pk) => pk,
-				None => return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
-			};
+					let signing_payload = stmt.signing_payload();
+					if !signature.verify(&signing_payload[..], &validator_public) {
+						return InvalidTransaction::BadProof.into();
+					}
 
-			let signing_payload = stmt.signing_payload();
-			if !signature.verify(&signing_payload[..], &validator_public) {
-				return InvalidTransaction::BadProof.into()
-			}
+					let active_vote = match PvfActiveVoteMap::<T>::get(&stmt.subject) {
+						Some(v) => v,
+						None => return InvalidTransaction::Custom(INVALID_TX_BAD_SUBJECT).into(),
+					};
 
-			let active_vote = match PvfActiveVoteMap::<T>::get(&stmt.subject) {
-				Some(v) => v,
-				None => return InvalidTransaction::Custom(INVALID_TX_BAD_SUBJECT).into(),
-			};
+					match active_vote.has_vote(validator_index) {
+						Some(false) => (),
+						Some(true) =>
+							return InvalidTransaction::Custom(INVALID_TX_DOUBLE_VOTE).into(),
+						None =>
+							return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
+					}
 
-			match active_vote.has_vote(validator_index) {
-				Some(false) => (),
-				Some(true) => return InvalidTransaction::Custom(INVALID_TX_DOUBLE_VOTE).into(),
-				None => return InvalidTransaction::Custom(INVALID_TX_BAD_VALIDATOR_IDX).into(),
+					ValidTransaction::with_tag_prefix("PvfPreCheckingVote")
+						.priority(T::UnsignedPriority::get())
+						.longevity(
+							TryInto::<u64>::try_into(
+								T::NextSessionRotation::average_session_length() / 2u32.into(),
+							)
+							.unwrap_or(64_u64),
+						)
+						.and_provides((stmt.session_index, stmt.validator_index, stmt.subject))
+						.propagate(true)
+						.build()
+				},
+				Call::apply_authorized_force_set_current_code { para, new_code } => {
+					match Self::validate_code_is_authorized(new_code, para) {
```

Comment:
Why do we have custom tx-validation? 

also, if this is going to be called via XCM, it will never go to the txpool anyways.

Reply:
> Why do we have custom tx-validation?
> 
> also, if this is going to be called via XCM, it will never go to the txpool anyways.

No, this is for a free transaction, and it will never be sent over XCM - that's the whole point of this PR: *"do not send the WASM code blob over XCM."*  

This tx validation is about `apply_authorized_force_set_current_code(code)`, not `pallet_xcm::send(authorize_force_set_current_code_hash(code_hash))`.